clippy_lints/src/transmute.rs

```diff
@@ -216,6 +216,31 @@ declare_clippy_lint! {
     "transmutes from a pointer to a pointer / a reference to a reference"
 }
 
+declare_clippy_lint! {
+    /// **What it does:** Checks for transmutes between collections whose it
+    /// types have different ABI, size or alignment.
+    ///
+    /// **Why is this bad?** This is undefined behavior.
+    ///
+    /// **Known problems:** Currently, we cannot know whether a type is a
+    /// collection, so we just lint the ones that come with `std`.
+    ///
+    /// **Example:**
+    /// ```rust
+    /// // different size, therefore likely out-of-bounds memory access
+    /// // You absolutely do not want this in your code!
+    /// std::mem::transmute<_, Vec<u32>>(vec![2_u16]);
+    /// ```
+    ///
+    /// You must always iterate, map and collect the values:
+    ///
+    /// ```rust
+    /// vec![2_u16].iter().map(u32::from).collect::<Vec<_>>()
```

Comment:
I'm not sure this is correct but the semicolon is needed at least.

```suggestion
    /// vec![2_u16].iter().map(|n| u32::from(*n)).collect::<Vec<_>>();
```